[macros] Introduce 'term.*' labels

jensmaurer opened the pull request.

jensmaurer committed:
[macros] Introduce 'term.*' labels

jensmaurer commented:
This still has

  Label `term.array.type' multiply defined
  Label `term.constant.expression' multiply defined
  Label `term.corresponding.object.parameter' multiply defined
  Label `term.more.specialized' multiply defined
  Label `term.multi-pass.guarantee' multiply defined
  Label `term.name' multiply defined
  Label `term.native.encoding' multiply defined
  Label `term.object.expression' multiply defined
  Label `term.parent.directory' multiply defined
  Label `term.ready' multiply defined
  Label `term.shared.lock' multiply defined
  Label `term.unevaluated.operand' multiply defined
  Label `term.union' multiply defined
  Label `term.usable.in.constant.expressions' multiply defined
  Label `term.width' multiply defined
Latexmk: Summary of warnings from last run of *latex:
  Latex found 16 multiply defined reference(s)

jensmaurer committed:
[std] Fix multiply-defined term.* labels by creating unique index entries

jensmaurer committed:
[check] Fail on multiply defined labels

jensmaurer committed:
[macros] Add \termref

jensmaurer committed:
[std] Use \termref throughout

jensmaurer commented:
@tkoeppe, this introduces \termref to refer to a defined term. Furthermore, all `\defn` terms automatically get a `term.*` label assigned. As an added bonus, you get a PDF hyperref for `\termref`:

![termref](https://user-images.githubusercontent.com/23412755/150017771-d8601374-3082-4258-adb6-cd532cc76689.png)

tkoeppe commented:
Is it the future already?!

tkoeppe commented on source/basic.tex:
```diff
@@ -5061,7 +5062,7 @@
 
 \item
-\defnx{unions}{\idxcode{union}}, which are classes capable of containing objects of
+\textit{unions}, which are classes capable of containing objects of
```
Why a bare `\textit` in the main document?

jensmaurer replied:
Because everything else in this itemized list is italized, and other bullets claim to contain definitions, but this one (and likely the other bullets, too) are not actually definitions. I wanted to keep the local italics consistency. If you think otherwise, I'll just remove the italics here.

tkoeppe replied:
Maybe use `\emph` then?

jensmaurer replied:
Fixed.

jensmaurer commented:
@Eelis , this might also interest your HTML conversion if/when it lands.

jensmaurer commented:
Editorial meeting 2022-01-21: Generally in favor. To be applied after the clause reorganization. Also globally replace `\defn` with `\termdef` (also related macros). The latter makes the change global and conflict-prone, so needs close coordination with @tkoeppe.

jensmaurer commented on source/basic.tex:
```diff
@@ -5068,7 +5069,7 @@
 on the access to these entities\iref{class.access};
 
 \item
-\defnx{unions}{\idxcode{union}}, which are classes capable of containing objects of
+\emph{unions}, which are classes capable of containing objects of
```
Yes. Most of the stuff here is not really a definition and should be cleaned up eventually. The current stage just removes the one conflict that was flagged by LaTeX.

jensmaurer commented on source/templates.tex:
```diff
@@ -3367,9 +3367,8 @@
-\indextext{more specialized!class template}%
 For two partial specializations,
-the first is \defn{more specialized} than the second if, given the following
+the first is \defnx{more specialized}{more specialized!class template} than the second if, given the following
```
So, moving that under "class template" is not good; we should keep it generic, right?

jensmaurer replied:
Function templates can't be partially specialized.

opensdh replied:
The index entry already said `!class template` because the whole partial-specialization section used to assume it was a class template.  I'm just pointing out that a line you're touching is (and already was) outdated.

Function templates can be more specialized, but they're not relevant here.

jensmaurer replied:
Fixed

jensmaurer committed:
[temp.spec.partial.order] Fix index entry: 'more specialized'

The phrase 'more specialized' also applies to variable
templates, not just to class templates.

AlisdairM commented:
Is this a useful update to our macros that we would want to try to complete and land for C++26?  Something to consider at the editors' meeting pending for Hagenberg?

AlisdairM commented:
As that the draft will be relatively stable during the ballot comment phase, with no significant new features to add, would this be a good time to update and land this PR that touches many places?  Assuming Jens would have the bandwidth to update the PR with everything else we will be throwing his way!
